sdk/core/core/src/az_url.c

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "az_hex_private.h"
+#include <az_precondition_internal.h>
+#include <az_url_internal.h>
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <_az_cfg.h>
+
+AZ_NODISCARD AZ_INLINE bool _az_url_should_encode(uint8_t c)
+{
+  switch (c)
+  {
+    case '-':
+    case '_':
+    case '.':
+    case '~':
+      return false;
+    default:
+      return !(('0' <= c && c <= '9') || ('A' <= c && c <= 'Z') || ('a' <= c && c <= 'z'));
+  }
+}
+
+AZ_NODISCARD az_result _az_url_encode(az_span destination, az_span source, int32_t* out_length)
+{
+  AZ_PRECONDITION_NOT_NULL(out_length);
+  AZ_PRECONDITION_VALID_SPAN(source, 0, true);
+
+  int32_t const input_size = az_span_size(source);
+  AZ_PRECONDITION_VALID_SPAN(destination, input_size, true);
```

Comment:
```suggestion
  AZ_PRECONDITION_VALID_SPAN(destination, input_size, false);
```

Reply:
Same as above.
`encode({ 0 }, { 0 }, &result)` is ok.

Reply:
I don't know if it should be. Why do you want to support this case?

Reply:
I don't want my code to care whether the spaniter pointer is NULL or not when span size is 0.
I find it convenient to use { 0 } spans.
And it is true, { 0 } is an empty span. A result of URL enconding an empty span is an empty span, no problem.
https://github.com/Azure/azure-sdk-for-c/pull/642

Reply:
Per our earlier online conversation, I'll update another PR to allow null input spans, but disallow null output spans in preconditions.